go-peer/chatroom.go

```diff
@@ -98,6 +124,73 @@ func (cr *ChatRoom) readLoop() {
 	}
 }
 
-func topicName(roomName string) string {
+// readFileLoop pulls messages from the pubsub file topic and handles them.
+func (cr *ChatRoom) readFileLoop() {
+	for {
+		msg, err := cr.fileSub.Next(cr.ctx)
+		if err != nil {
+			close(cr.Messages)
+			return
+		}
+		// only forward messages delivered by others
+		if msg.ReceivedFrom == cr.h.ID() {
+			continue
+		}
+
+		fileID := msg.Data
+		fileBody, err := cr.requestFile(msg.GetFrom(), fileID)
+		if err != nil {
+			close(cr.Messages)
+			return
+		}
+
+		cm := new(ChatMessage)
+		cm.Message = fmt.Sprintf("File: %s (%v bytes) from %s", string(fileID), len(fileBody), msg.GetFrom().String())
+		cm.SenderID = msg.ID
+		cm.SenderNick = string(msg.ID[len(msg.ID)-8])
+		// send valid messages onto the Messages channel
+		cr.Messages <- cm
+	}
+}
+
+func (cr *ChatRoom) requestFile(toPeer peer.ID, fileID []byte) ([]byte, error) {
+	stream, err := cr.h.NewStream(context.Background(), toPeer, "/universal-connectivity-file/1")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create stream: %w", err)
+	}
+	defer stream.Close()
+
+	reqLen := binary.AppendUvarint([]byte{}, uint64(len(fileID)))
+	if _, err := stream.Write(reqLen); err != nil {
+		return nil, fmt.Errorf("failed to write fileID to the stream: %w", err)
+	}
+	if _, err := stream.Write(fileID); err != nil {
+		return nil, fmt.Errorf("failed to write fileID to the stream: %w", err)
+	}
+	if err := stream.CloseWrite(); err != nil {
+		return nil, fmt.Errorf("failed to close write stream: %w", err)
+	}
+
+	reader := bufio.NewReader(stream)
+	respLen, err := binary.ReadUvarint(reader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response length prefix: %w", err)
+	}
+	fileBody := make([]byte, respLen)
+	if _, err := reader.Read(fileBody); err != nil {
+		return nil, fmt.Errorf("failed to read fileBody from the stream: %w", err)
+	}
+	if err := stream.CloseRead(); err != nil {
+		return nil, fmt.Errorf("failed to close read stream: %w", err)
+	}
+
+	return fileBody, nil
+}
+
+func chatTopicName(roomName string) string {
 	return roomName
 }
+
```

Comment:
```suggestion

// fileTopicName returns the name of the pubsub topic used for sending/recieving files in the chat room.
```

Reply:
Thank you, @TheDiscordian! Just applied all comments to be added.